File: app/src/main/java/com/mx/gillustrated/adapter/SpinnerCommonAdapter.java
package com.mx.gillustrated.adapter;

import android.content.Context;
import android.os.Build;
import android.text.Html;
import android.text.Spanned;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import com.mx.gillustrated.vo.CardInfo;
import com.mx.gillustrated.vo.EventInfo;
import com.mx.gillustrated.vo.SpinnerInfo;

import java.util.List;

/**
 * Created by maoxin on 2018/7/17.
 */

public class SpinnerCommonAdapter<T extends SpinnerInfo> extends BaseAdapter {

    private Context mcontext;
    private LayoutInflater layoutInflater;
    private static int mResource = android.R.layout.simple_spinner_item;
    private static int mDropDownResource = android.R.layout.simple_spinner_dropdown_item;
    private List<T> list;
    private boolean mExtraParentheses = true;
    private int mFontSize = 0;

    public SpinnerCommonAdapter(Context context, List<T> items) {
        mcontext = context;
        layoutInflater = LayoutInflater.from(mcontext);
        list = items;
    }

    public SpinnerCommonAdapter(Context context, List<T> items, boolean extraParentheses, int fontSize) {
        mcontext = context;
        layoutInflater = LayoutInflater.from(mcontext);
        list = items;
        this.mExtraParentheses = extraParentheses;
        this.mFontSize = fontSize;
    }


    @Override
    public int getCount() {
        return list.size();
    }

    @Override
    public T getItem(int position) {
        return list.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    //Dropdown Item Style
    @Override
    public View getDropDownView(int position, View convertView, ViewGroup parent) {
        return createViewFromResource(position, convertView, parent, mDropDownResource, false );
    }

    //Select Item Style
    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        return createViewFromResource(position, convertView, parent, mResource, true);
    }

    private View createViewFromResource(int position, View convertView,
                                        ViewGroup parent, int resource, boolean defaultView) {
        View view;
        TextView text;

        if (convertView == null) {
            view = layoutInflater.inflate(resource, parent, false);
        } else {
            view = convertView;
        }
        text = (TextView) view;
        if(mFontSize > 0){
            text.setTextSize(mFontSize);
            if(defaultView)
                text.setPadding(0,0,0,0);
        }

        T items = getItem(position);

        String textToShow = mExtraParentheses && items.getNid() > 0
                ? items.getName() + " (" + items.getNid() + ")" : items.getName();
        Spanned spanned = Build.VERSION.SDK_INT >= Build.VERSION_CODES.N
                ? Html.fromHtml(textToShow, Html.FROM_HTML_MODE_LEGACY)
                : Html.fromHtml(textToShow);
        text.setText(spanned, TextView.BufferType.SPANNABLE);

        return view;

    }
}

File: app/src/main/java/com/mx/gillustrated/adapter/SpinnerCommonAdapter.kt
package com.mx.gillustrated.adapter

import android.content.Context
import android.os.Build
import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.TextView
import com.mx.gillustrated.vo.SpinnerInfo

/**
 * Created by maoxin on 2018/7/17.
 */

class SpinnerCommonAdapter<T : SpinnerInfo> constructor(mContext: Context, private val list: List<T>) : BaseAdapter() {

    private var layoutInflater: LayoutInflater = LayoutInflater.from(mContext)
    private var mExtraParentheses = true
    private var mFontSize = 0

    constructor(context: Context, items: List<T>, extraParentheses: Boolean, fontSize: Int): this(context, items) {
        this.mExtraParentheses = extraParentheses
        this.mFontSize = fontSize
    }


    override fun getCount(): Int {
        return list.size
    }

    override fun getItem(position: Int): T {
        return list[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    //Dropdown Item Style
    override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
        return createViewFromResource(position, convertView, parent, mDropDownResource, false)
    }

    //Select Item Style
    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        return createViewFromResource(position, convertView, parent, mResource, true)
    }

    private fun createViewFromResource(position: Int, convertView: View?,
                                       parent: ViewGroup, resource: Int, defaultView: Boolean): View {

        val view: View = convertView ?: layoutInflater.inflate(resource, parent, false)
        val text: TextView

        text = view as TextView
        if (mFontSize > 0) {
            text.textSize = mFontSize.toFloat()
            if (defaultView)
                text.setPadding(0, 0, 0, 0)
        }

        val items = getItem(position)

        val textToShow = if (mExtraParentheses && items.nid > 0)
            items.name + " (" + items.nid + ")"
        else
            items.name
        val spanned = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N)
            Html.fromHtml(textToShow, Html.FROM_HTML_MODE_LEGACY)
        else
            Html.fromHtml(textToShow)
        text.setText(spanned, TextView.BufferType.SPANNABLE)

        return view

    }

    companion object {
        private const val mResource = android.R.layout.simple_spinner_item
        private const val mDropDownResource = android.R.layout.simple_spinner_dropdown_item
    }
}
